Les algorithmes gloutons
=====================

## Introduction
Dans certains problèmes appelés *problèmes d'optimisation* on doit maximiser ou minimiser une fonction en tenant compte d'une ou de plusieurs contraintes. Les algorithmes utilisés font intervenir une série d'étapes, chaque étape proposant un ensemble de choix. 

L'approche naïve consistant à énumérer tous les choix possibles (*brute force*) est rarement utilisable à cause de son coût.  

L'approche gloutonne, qui sans être nécessairement optimale, est assez rapide et simple à mettre en oeuvre. Il s'agit de **faire le choix qui semble le meilleur à chaque étape** (en espérant que celui-ci mène à une solution optimale) et ne plus revenir sur ce choix. 

Deux exemples sont présentés en première: le problème du rendu de monnaie et le problème du sac à dos.

## Le problème du rendu de monnaie
Le problème du rendu de monnaie s'énonce de la façon suivante : étant donné un système de monnaie (*pièces et billets, qu'on confondra ici pour simplifier, exceptés les centimes*), comment rendre une somme donnée de façon optimale, c'est-à-dire avec le nombre minimal de pièces et/ou billets ?

**Un exemple**: le boulanger doit rendre la monnaie de 9€. 
1. Donner toutes les combinaisons possibles ainsi que le nombre de pièces utilisées. 
2. Quelle stratégie adopter pour avoir la solution optimale?

L'algorithme glouton utilisé ici peut s'énoncer comme suit:

```
fonction rendu_monnaie(montant,monnaie)
//montant: entier
//monnaie: tableau d'entier correspondant aux valeurs du système monétaire utilisé
//trié par ordre croissant

Affecter à une variable n la taille du tableau monnaie
Initialiser un tableau 'solution' à zéro   //correspond aux pièces utilisées
Pour i allant de n-1 à 0
    Tant que montant >= monnaie[i] 
        montant <- montant - monnaie[i]
        solution[i] <- solution[i] + 1
Retourner le tableau solution
```

3. Coder cet algorithme en python.
4. Quelle est la complexité en temps de cet algorithme ?

In [1]:
def rendu_monnaie(montant, monnaie):
    """
    Retourne un tableau d'entiers indiquant le nombre de type de pièces prises;
    montant: entier positif
    monnaie: tableau d'entiers positif représentant le système monétaire utilisé.
    """
    n = len(monnaie)
    #Votre code à partir d'ici

Pour tester votre fonction, ajouter une cellule avec le contenu ci-après et exécuter.
```python
euros = [1, 2, 5, 10, 20, 50, 100]
assert rendu_monnaie(37, euros) == [0, 1, 1, 1, 1, 0, 0]
```

5. Soit le système monétaire formé des valeurs suivantes:
```python
monnaie1 = [1, 6, 10]
```
Appeler la fonction `rendu_monnaie(12, monnaie1)`.
Commenter.

6. Soit le système monétaire formé des valeurs suivantes:
```python
monnaie2 = [2, 3]
```
Que se passerait-il si on appelait (**ne pas le faire**) `rendu_monnaie(4, monnaie2)`. Commenter

Les euros forment un système *canonique*, c'est-à-dire un système pour lequel l'algorithme glouton donne une solution optimale.

## Le problème du sac à dos

Le problème consiste à remplir un sac à dos ayant une capacité maximale fixée, avec tout ou partie d'un ensemble donné d'objets ayant chacun une masse et une valeur. Les objets mis dans le sac à dos doivent maximiser la valeur totale, sans dépasser la capacité (*masse*) maximale supportée par le sac.

### S'approprier le problème
Télécharger le fichier [sac-a-dos.zip](https://brdarid.nohost.me/nextcloud/s/WHKozDrAoi44pG3), le décompresser puis, dans le dossier `sac-a-dos` obtenu, ouvrir `applet.html` avec un navigateur. Tester !

### Vers le codage d'un algorithme en python
Comme cela a été précisé en introduction, appliquer un algorithme glouton revient à faire le **meilleur choix**  à chaque étape. Comment définir ce meilleur choix ? ou, en d'autres termes, quel objet prendre ?
Trois options:  
   * valeur maximale;
   * masse minimale;
   * rapport (valeur / masse) maximal  

La structure de données envisagée pour modéliser les objets: tableau de p-uplets nommés (*en 1ere NSI des dictionnaires*) ayant pour clés un index, la valeur et la masse).  
On donne ci-dessous un code commenté si le choix se fait suivant la valeur (*ordre décroissant*), ainsi qu'un exemple d'appel de la fonction construite.

In [18]:
def valeur(objet):
    return objet['valeur']

def glouton(tab_objet, capacite, choix):
    """
    Renvoie un p-uplet formé d'un tableau de solution (0 ou 1 suivant que l'objet est pris ou pas) 
    et de la valeur totale des objets pris.
    tab_objet: tableau de p-uplets nommés (dictionnaire), chacun représentant un objet avec les 
    clés 'index', 'valeur' et 'masse';
    capacite: entier, contrainte (ici masse maxi)
    choix: fonction (utile au tri des objet, voir ci-dessus)
    """    
    vtotale = 0
    mtotale = 0
    objets_pris = [0] * len(tab_objet)
    #Tri des objets (critère de choix à indiquer avec le paramètre key)
    tab_objet_trie = sorted(tab_objet, key=choix, reverse=True)
    #Parcours de la liste d'objets et les ajouter à objets_pris SI la contrainte 
    #est respectée, c-a-d masse totale <= capacité
    for i in range(len(tab_objet)):
        if mtotale + tab_objet_trie[i]['masse'] <= capacite:
            objets_pris[tab_objet_trie[i]['index']] = 1
            vtotale += tab_objet_trie[i]['valeur']
            mtotale += tab_objet_trie[i]['masse']
    return objets_pris, vtotale
            

In [19]:
obj = [{'index': 0, 'valeur': 126, 'masse': 14}, 
      {'index': 1, 'valeur': 32, 'masse': 2}, 
      {'index': 2, 'valeur': 20, 'masse': 5}, 
      {'index': 3, 'valeur': 5, 'masse': 1}, 
      {'index': 4, 'valeur': 18, 'masse': 6}, 
      {'index': 5, 'valeur': 80, 'masse': 8}]

In [20]:
glouton(obj, 15, valeur)

([1, 0, 0, 1, 0, 0], 131)

La réponse ci-dessus s'interprète de la manière suivante: les objets 0 et 3 sont pris; la valeur des objets pris est de 131.

#### Travail 

1. A quel endroit du programme, la stratégie gloutonne (*faire le choix qui semble le meilleur à chaque étape et ne plus revenir sur ce choix*) est bien mise en évidence ?

2. Proposer les modifications nécessaires de manière à ce que le choix d'un objet se fasse selon:  
    * la masse décroissante:
    * le rapport $\dfrac{\mathrm{valeur}}{\mathrm{masse}}$ décroissant.